components/socket/net/index.mjs

```diff
@@ -0,0 +1,70 @@
+import { connect } from "net";
+import { fileURLToPath } from "url";
+import { Buffer } from "buffer";
+import NetSocketMessaging from "net-socket-messaging";
+
+const { concat: concatBuffer } = Buffer;
+const { createMessage } = NetSocketMessaging;
+
+export default (dependencies) => {
+  const {
+    path: { toIPCPath },
+    util: { mapMaybe },
+    log: { logGuardWarning, logWarning },
+  } = dependencies;
+  const flush = (socket, messages) => {
+    if (socket.writable && messages.length > 0) {
+      const buffers = messages.map(createMessage);
+      // Socket.write registers asynchronous resources hence synchronous
+      // message writtting so we need to empty the messages array before
+      messages.length = 0;
+      socket.write(concatBuffer(buffers));
+    }
+  };
+  return {
+    openSocket: (host, port, { heartbeat, threshold }) => {
+      const socket =
+        typeof port === "string"
+          ? connect(toIPCPath(fileURLToPath(port)))
+          : connect(port, host);
+      socket.unref();
+      const messages = [];
+      socket.on("connect", () => {
+        const flushBind = () => {
+          flush(socket, messages);
+        };
+        process.once("beforeExit", flushBind);
+        flushBind();
+        const timer = mapMaybe(heartbeat, () =>
+          setInterval(flushBind, heartbeat),
+        );
+        mapMaybe(timer, (timer) => {
+          timer.unref();
+        });
+        socket.on("close", () => {
+          process.off("beforeExit", flushBind);
+          logGuardWarning(messages.length > 0, "Lost messages >> %j", messages);
+          messages.length = 0;
+          clearInterval(timer);
+        });
+      });
+      return { messages, socket, threshold };
+    },
+    closeSocket: ({ socket, messages }) => {
+      // TODO: investigate how to make this flush is done synchronously
+      // because closeSocket is often called on process exit event.
```

Comment:
The data in the internal buffer? In other components, `closeSocket` is invoked on the process `exit` event. The handler of this event can only do synchronous work before the process truly exit. Because `net.Socket.prototype.write` is actually an asynchronous operation, it is not guaranteed that the data will reach the other end point.

Reply:
Well, yes, but what's going to be in the internal buffer? ;) 

I haven't really dug in to the inner workings of the agent. So, I don't know what's going to be writing to a socket, or what data is going to be in the internal buffer at process exit.

I think if you could give an example of a flow that would lead to data loss, that would help me understand better.

Reply:
Basically, every data that the client is collecting about the recorded application is send to the server for post-processing via this socket. So basically: file information for populating `appmap.classMap` and function applications for populating `appmap.events`.

For instance of everything happens synchronously during the initial event processing:
```js
function main () {};
main();
// Process exit happens here because there nothing else to do
// But internal buffer is populated with:
//   - information about this file
//   - information about main function application 
```

Reply:
Ok. So, if the client exits while a recording is in progress, some of the events could be lost. Does the server handle the missing data gracefully?

Reply:
Yes trace completion is manufactured on the server. So the client can crash at any point and the already received data are guaranteed to be part of the appmap.

Reply:
For completeness. This problem only happens when fallback to `net.Socket`. Also, I'm listening to `beforeExit` event so that I have the change to perform asynchronous operation. But this happens only once to avoid infinite loops and prevent process exit altogether. So if the recorded application also has `beforeExit` handlers, it can revive the process (which is hard to detect) and we loose the chance to perform asynchronous operation and rely on the synchronous `exit` handler.

Reply:
In short, losing data of interest should remain marginal and when it happens the server can still make sense of whatever data it already received.